relayer/relays/beacon/header/header.go

```diff
@@ -82,23 +83,37 @@ func (h *Header) Sync(ctx context.Context, eg *errgroup.Group) (<-chan uint64, <
 	basicChannel := make(chan uint64)
 	incentivizedChannel := make(chan uint64)
 
+	lastSyncedExecutionBlockNumber := uint64(0)
+
 	eg.Go(func() error {
 		defer func() {
 			close(basicChannel)
 			close(incentivizedChannel)
 		}()
 		for {
-			lastBlockNumber, err := h.SyncHeaders(ctx)
+			err := h.SyncHeaders(ctx)
 			switch {
 			case errors.Is(err, ErrFinalizedHeaderUnchanged):
 				log.WithField("finalized_header", h.cache.LastFinalizedHeader()).Info("finalized header unchanged")
+			case errors.Is(err, ErrFinalizedHeaderNotImported):
+				log.Warn("last finalized header wasn't imported")
 			case err != nil:
 				log.WithError(err).Error("error while syncing headers")
-
 				return err
 			default:
-				basicChannel <- lastBlockNumber
-				incentivizedChannel <- lastBlockNumber
+				executionBlockNumber, err := h.syncer.GetExecutionBlockHash(h.cache.LastFinalizedHeader())
+				if err != nil {
+					return fmt.Errorf("fetch execution block hash: %w", err)
+				}
+
+				if executionBlockNumber > lastSyncedExecutionBlockNumber {
+					lastSyncedExecutionBlockNumber = executionBlockNumber
+
+					log.WithField("block_number", lastSyncedExecutionBlockNumber).Info("sending block number")
+
+					basicChannel <- lastSyncedExecutionBlockNumber
+					incentivizedChannel <- lastSyncedExecutionBlockNumber
```

Comment:
This code here can deadlock if the receiving goroutine has died and is no longer reading from the channel.

Solution is to use a select statement that waits until an item is written or when the context has been cancelled.
```golang
select {
case <-ctx.Done():
	return ctx.Err()
case someChannel <- item:
}
```

Reply:
I added the select just a little bit further down: https://github.com/Snowfork/snowbridge/blob/6603d1d3e73d15cbd2a3a605ea16d22d8dd1e585/relayer/relays/beacon/header/header.go#L119 Let me know if you think it will still be an issue, but as far as I know it does check if the context has been cancelled.

The reason why the select statement is below the other code is so that the code block executes immediately, instead of waiting for the timer. https://github.com/Snowfork/snowbridge/pull/668#discussion_r946479917

Reply:
It will still be an issue, as the problem I'm describing isn't really related to the ticker/timer. The channel write statements on lines 113 and 114 could block forever (i.e deadlock). 

In Go, reading and writing from a channel happens synchronously, even if the reader and writer are different goroutines. Lets consider the `basicChannel` channel. Some other goroutine is reading from it. lets call it goroutine A. The goroutine writing to it, lets call it goroutine B. 

If goroutine A dies while goroutine B is trying to write to the channel, then goroutine B will block _forever_ trying to write to the channel. To prevent this, we need some signalling mechanism, so that goroutine A can tell goroutine B that it dying and that it should stop trying to write to the channel. That signalling mechanism is implemented using the Errgroup/Context structures we pass around everywhere.

In Golang, a blocking channel write looks like this. Nothing can interrupt this write while it is blocking:
```golang
someChannel <- item
```

We can implement a non-blocking channel write using select:
```
select {
case someChannel <- item:
default:
   log.Info("Cannot write to channel right now")
}

# code continues executing
...
```

Now in our case, we want to wait until we can write to `basicChannel` and `incentivizedChannel`  OR return immediately if the relayer is about to self-terminate
```
select {
case basicChannel <- item:
case <-ctx.Done():
	return ctx.Err()
}

select {
case incentivizedChannel <- item:
case <-ctx.Done():
	return ctx.Err()
}
```

@alistair-singh @doubledup ^ also useful for you to read through this

Reply:
Thanks for explaining @vgeddes! Let me change it.

Reply:
Addressed in https://github.com/Snowfork/snowbridge/pull/681/commits/fd73a7deedf1168bac09174f122ef300f36d8a7f.